### PR TITLE
gui: Add default ignores to the advanced configuration editor modal (fixes #8264)

### LINKED ISSUE #8264: Include default ignore patterns in the GUI's advanced configuration editor
The new configuration option added in #7530 can be set through the Edit Folder Defaults button in the GUI's regular Settings dialog.  But it's missing in the Advanced Configuration modal.

### PR DESCRIPTION
**Pull request by acolomb**
### Purpose

Fixes #8264.

Follow the XML / JSON structure and keep it under a separate headline instead of the "default folder" section (as in the regular settings modal).

This section slightly differs in the handling of list-type config options.  Usually they are represented as comma-separated single-line text inputs, which is unsuitable for ignore patterns which may rightfully contain commas.  Thus a textarea is used, using newline as the list delimiter.

### Testing

Verified that patterns entered under Edit Folder Defaults are correctly adapted in Advanced Configuration and vice versa.  Leading and trailing spaces are preserved in both directions.

### Screenshots

image

### CODE REVIEW

**Comment by tomasz1986** [comment]
I'm not sure what is responsible for this, but empty lines seem to get trimmed on save. I don't believe this is the desired behaviour (as users will want to have empty lines in their ignore patterns, e.g. to separate different sections). Empty lines are preserved when using the "non-advanced" default ignore patterns editor.

**Comment by acolomb** [comment]
We do already set `ng-trim="false"`, so I'm not sure what else could be done about that. It uses the JSON interface to manipulate the config directly, while the other, human-friendly editor uses a different endpoint. As this function is really just for completeness sake, I guess whoever uses it may accept this slight nuisance? After all they're just the default patterns and functionally identical without the blank lines. Copying a pretty version over the one created from defaults helps as a workaround I guess, without hurting the actual syncing.

Curious as I haven't tested: How does the CLI operation deal with this?

**Comment by tomasz1986** [comment]
> We do already set `ng-trim="false"`

Shouldn't it actually trim white spaces though? The human-friendly editor trims both leading and trailing spaces, and this is done by design, as I was told in one of the old issues (which I cannot find now), where I asked about this behaviour.

**Comment by acolomb** [comment]
I cannot confirm that the other settings editor trims spaces.  It may happen when actually submitting the ignore patterns on an existing folder, but not for the defaults.  I entered patterns with leading and trailing spaces in the "Edit Folder Defaults" modal and they ended up verbatim in the config file.

**Comment by tomasz1986** [comment]
It does trim, but after more testing, it doesn't seem to work perfectly.

For example, when starting from a clean state, if you input " abc" (not using backticks here, because GitHub itself hides spaces), it will trim the leading space. On the other hand, if you input "abc" and then " def" in the second line, the leading space will be preserved.

I've found the original issue I was referring to before, which is https://github.com/syncthing/syncthing/issues/7228. The intended behaviour is to trim both leading and trailing white space.

**Comment by acolomb** [comment]
There is not much we can do about it.  I first tried *not* using `ng-trim="false"` and then every character ended up on its own line.  I don't think this is worth any effort digging into preserving or removing unnecessary whitespace right in this function.  It's just an oversight that the advanced config editor is missing.  The value will get sanitized before actually using it, e.g. copying to an `.stignore` file.  Might also get sanitized after reading from there, but the point is, we don't need to do it here in the "raw" config editor.

**Comment by imsodin** [comment]
Does that mean that if I have default ignores with formatting like empty lines, I open advanced settings to change something entirely different, save and then my default ignores change? In that case I'd rather miss default ignores in advanced settings than having this behaviour.

**Comment by acolomb** [comment]
Yes, it will drop blank lines just by saving from the advanced settings modal.  The only hack around that coming to my mind is artificially adding a space to every blank line before it's copied to the textarea.  Or we need custom handling of the multi-line text to array of lines conversion.  Angular's `ng-list` directive seems too unflexible for this.

**Comment by acolomb** [comment]
> but I'd just like to point out that the "normal", i.e. per-folder ignore patterns, are also currently missing from there 😉.

Of course they are. They're not even stored in the config.

I will investigate some more, but give me a couple of days.

**Comment by acolomb** [comment]
Now I avoided the `ng-list` directive completely and added a special case for the `lines` field using a getter / setter function.  See https://docs.angularjs.org/api/ng/directive/ngModel under the *Binding to a getter/setter* headline.  During my testing, this yielded an accurate reproduction of the entered patterns including the blank lines and any leading / trailing whitespace.

File: gui/default/syncthing/core/syncthingController.js [hunk 1]
```diff
@@ -2777,6 +2777,12 @@ angular.module('syncthing.core')
             $scope.advancedConfig = angular.copy($scope.config);
             $scope.advancedConfig.devices.sort(deviceCompare);
             $scope.advancedConfig.folders.sort(folderCompare);
+            $scope.advancedConfig.defaults.ignores._lines = function (newValue) {
+                if (arguments.length) {
```
Comment (imsodin) [inline gui/default/syncthing/core/syncthingController.js:2781]: `newValue.length`?

File: gui/default/syncthing/settings/advancedSettingsModalView.html [hunk 2]
```diff
@@ -176,8 +176,12 @@ <h4 class="panel-title" tabindex="0" translate>Default Ignore Patterns</h4>
...
-                      <textarea ng-if="inputTypeFor(key, value) == 'list'" id="advancedDefaultIgnoresInput{{$index}}" class="form-control" ng-model="advancedConfig.defaults.ignores[key]" ng-list="&#10;" ng-trim="false"></textarea>
-                      <input ng-if="inputTypeFor(key, value) != 'list'" id="advancedDefaultIgnoresInput{{$index}}" class="form-control" type="{{inputTypeFor(key, value)}}" ng-model="advancedConfig.defaults.ignores[key]" />
+		      <div ng-switch="key">
+			<!-- Special case to preserve empty lines in multi-line input -->
+			<textarea ng-switch-when="lines" id="advancedDefaultIgnoresInput{{$index}}" class="form-control" ng-model="advancedConfig.defaults.ignores._lines" ng-model-options="{ getterSetter: true }"></textarea>
+			<input ng-switch-default ng-if="inputTypeFor(key, value) == 'list'" id="advancedDefaultIgnoresInput{{$index}}" class="form-control" type="text" ng-model="advancedConfig.defaults.ignores[key]" ng-list />
+			<input ng-switch-default ng-if="inputTypeFor(key, value) != 'list'" id="advancedDefaultIgnoresInput{{$index}}" class="form-control" type="{{inputTypeFor(key, value)}}" ng-model="advancedConfig.defaults.ignores[key]" />
```
Comment (imsodin) [inline gui/default/syncthing/settings/advancedSettingsModalView.html:183]: indentation

File: gui/default/syncthing/settings/advancedSettingsModalView.html [hunk 3]
```diff
@@ -167,6 +167,26 @@ <h4 class="panel-title" tabindex="0" translate>Default Device</h4>
                 </form>
               </div>
             </div>
+            <div class="panel panel-default">
+              <div class="panel-heading" role="tab" id="advancedDefaultIgnoresHeading" data-toggle="collapse" data-parent="#advancedDefaults" href="#advancedDefaultIgnores" aria-expanded="false" aria-controls="advancedDefaultIgnores" style="cursor: pointer;">
+                <h4 class="panel-title" tabindex="0" translate>Default Ignore Patterns</h4>
+              </div>
+              <div id="advancedDefaultIgnores" class="panel-collapse collapse" role="tabpanel" aria-labelledby="advancedDefaultIgnoresHeading">
+                <form class="form-horizontal" role="form">
+                  <div ng-repeat="(key, value) in advancedConfig.defaults.ignores" ng-if="inputTypeFor(key, value) != 'skip'" class="form-group">
```
Comment (imsodin) [inline gui/default/syncthing/settings/advancedSettingsModalView.html:176]: Is this for future extensibility? Afaik currently there's just a single `lines` key here, i.e. no looping and no `inputTypeFor` conditions below will apply.

File: gui/default/syncthing/core/syncthingController.js (hunk 1 above)
Comment (acolomb) [reply gui/default/syncthing/core/syncthingController.js:2781]: Doesn't work when the argument is `undefined` (omitted), which happens when it's called as a getter function. The technique is taken straight from from Angular's documentation.

File: gui/default/syncthing/settings/advancedSettingsModalView.html (hunk 3 above)
Comment (acolomb) [reply gui/default/syncthing/settings/advancedSettingsModalView.html:176]: Yes, I added it so we have the same generic mechanism for future additions as above without touching this code again. I could imagine some boolean options for when the patterns should apply for example.

File: gui/default/syncthing/core/syncthingController.js (hunk 1 above)
Comment (imsodin) [reply gui/default/syncthing/core/syncthingController.js:2781]: TIL that there's an `arguments` object in all JS functions.

File: gui/default/syncthing/settings/advancedSettingsModalView.html (hunk 2 above)
Comment (acolomb) [reply gui/default/syncthing/settings/advancedSettingsModalView.html]: Fixed, damn tabs setting.  Maybe I should add a `.editorconfig` locally...

**Review by imsodin — APPROVED** [review]
I haven't tested it, but looks good. Maybe @tomasz1986 wants to have a look?

**Comment by tomasz1986** [comment]
> I haven't tested it, but looks good. Maybe @tomasz1986 wants to have a look?

I've tested the current iteration just now. It seems to be working fine 🙂.

There're still some quirks regarding empty lines and spaces between the two editors, e.g. if you leave the very first line blank in the "advanced" editor, and then open the "normal" editor, it will automatically be trimmed there. I guess this doesn't really matter in the grand scheme of things, as long as empty lines between the actual ignore patterns are preserved.

I'd just like to suggest adding `rows="5"` to the `textarea` as the "normal" editor does, which will make the default size of the two identical.

**Comment by acolomb** [comment]
Five rows it is.

**Comment by tomasz1986** [comment]
Looks good to me. Do you need to run the translate script too, or will this be processed automatically later on?

**Comment by acolomb** [comment]
It will run as part of the GitHub -> Transifex import, but doesn't hurt to do it here anyway.  Note that in general you should be careful with adding stuff manually, because it won't be removed again if the strings change in the course of PR updates.

**Comment by acolomb** [comment]
@imsodin Are you okay with this now?

**Review by imsodin — APPROVED** [review]
I already was last time I reviewed, just wanted tomasz to have another look and then didn't act once he did.

**Pull request merged**